xdsl/dialects/pdl_interp.py

```diff
@@ -417,6 +422,48 @@ def __init__(
         super().__init__(operands=[lhs, rhs], successors=[trueDest, falseDest])
 
 
+@irdl_op_definition
+class ApplyConstraintOp(IRDLOperation):
+    """
+    See external [documentation](https://mlir.llvm.org/docs/Dialects/PDLInterpOps/#pdl_interpapply_constraint-pdl_interpapplyconstraintop).
+    """
+
+    name = "pdl_interp.apply_constraint"
+    traits = traits_def(IsTerminator())
+    constraint_name = prop_def(StringAttr, prop_name="name")
+    is_negated = prop_def(
+        BoolAttr, prop_name="isNegated", default_value=BoolAttr(False, i1)
+    )
+    args = var_operand_def(AnyPDLTypeConstr)
+    rresults = var_result_def(AnyPDLTypeConstr)
+    true_dest = successor_def()
+    false_dest = successor_def()
+    irdl_options = [ParsePropInAttrDict()]
+
+    assembly_format = "$name `(` $args `:` type($args) `)` `:` type($rresults) attr-dict `->` $true_dest `, ` $false_dest"
+
+    def __init__(
+        self,
+        constraint_name: str | StringAttr,
+        args: Sequence[SSAValue],
+        true_dest: Block,
+        false_dest: Block,
+        res_types: Sequence[AnyPDLType] = (),
+        is_negated: bool = False,
```

Comment:
```suggestion
        is_negated: bool | BoolAttr = False,
```
We tend to allow attributes to be passed in when we already have them.